### PR TITLE
Add Automated deployment of `main` branch to Development registry

### LINKED ISSUE #49: Create CD for deploying registry code to production
Currently the registry's IaC and code are being deployed locally. It's hard to follow who deployed what to production
We should have a GHA that runs those deployments, and only use that:

GHA should basically `opentf/terraform init` and `opentf/terraform apply`, with the correct configuration, with a configurable branch

### PR DESCRIPTION
**Pull request by eranelbaz**
Will partially help with #49
Afterwards, we'd need a CD process for production as well

basic CD process
currently using `terraform` as we yet to have `tofu` release

### CODE REVIEW

File: .github/workflows/continuous-deployment.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        uses: aws-actions/configure-aws-credentials@v4
+        with:
+          role-to-assume: arn:aws:iam::743518529013:role/GithubActions-AssumeRoleWithAction-Registry
+          aws-region: eu-west-1
+
+      - name: Run terraform
+        run: |
+          terraform init
+          terraform workspace select dev
+          terraform apply -var="github_api_token=${{ github.token }} " -var="route53_zone_id=Z08649001JBA38NN3K55" -var="domain_name=devregistry.opentofu.org" -auto-approve
```
Comment (eranelbaz) [inline .github/workflows/continuous-deployment.yml]: please double check the zone id and domain name

**Comment by RLRabinowitz** [comment]
I like the automatic deployments for the dev registry!

@eranelbaz WDYT about allowing this Action to trigger the deployment of the production registry manually with workflow_dispatch?
I think that was the main concern of [the issue](https://github.com/opentofu/registry/issues/49)

File: .github/workflows/continuous-deployment.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - uses: hashicorp/setup-terraform@v2
+        with:
+          terraform_version: 1.5.6
+
+      - name: Configure AWS Credentials
+        uses: aws-actions/configure-aws-credentials@v4 # https://github.com/aws-actions/configure-aws-credentials#sample-iam-oidc-cloudformation-template
```
Comment (RLRabinowitz) [inline .github/workflows/continuous-deployment.yml]: The comment is related to creation of the user, right?
I'm pretty sure it's not really necessary here, but you can move the comment to be next to the `role-to-assume` and write `# Role created following this guideline: https://github.com/aws-actions/configure-aws-credentials#sample-iam-oidc-cloudformation-template`

**Review by Yantrio — COMMENTED** [review]
The flow looks good.

A few small nitpicks:

1. Let's make things use variables.
	- zoneid, domainname, workspacename, github token, ARN for role
2. This should probably have a dependency on the test workflow, to make sure we don't deploy bad code.

3. Don't use the github token from the workflow process, we need to be using a specific token.

**Comment by Yantrio** [comment]
Before this is merged, we should introduce state locking. I'll make an issue for this now.

File: .github/workflows/continuous-deployment.yml (hunk 1 above)
Comment (Yantrio) [reply .github/workflows/continuous-deployment.yml]: Moved to a secret

File: .github/workflows/continuous-deployment.yml (hunk 2 above)
Comment (Yantrio) [reply .github/workflows/continuous-deployment.yml]: I moved this comment above so its a bit clearer

File: .github/workflows/dev-deployment.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
+name: Dev Deployment
+
+on:
+  push:
+    branches: [ main ]
+  workflow_dispatch:
```
Comment (cube2222) [inline .github/workflows/dev-deployment.yml:6]: Just making sure: Is the AWS role limited to only work on the main branch?

File: .github/workflows/dev-deployment.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+      - name: Select Terraform Workspace
+        if: success()
+        run: terraform workspace select dev
+        continue-on-error: false
+
+      - name: Run terraform
+        if: success()
+        run:
+          terraform apply -auto-approve
```
Comment (cube2222) [inline .github/workflows/dev-deployment.yml]: 😨 don't we want a separate planning step, or something like that?

Just `apply -auto-approve` without confirmation sounds really scary.

File: .github/workflows/dev-deployment.yml (hunk 3 above)
Comment (Yantrio) [reply .github/workflows/dev-deployment.yml:6]: Yes it is.

File: .github/workflows/dev-deployment.yml (hunk 4 above)
Comment (Yantrio) [reply .github/workflows/dev-deployment.yml]: I've introduced this fix in the latest branch

**Comment by eranelbaz** [comment]
@Yantrio thanks for the fixes on the pr as I'm away!

**Comment by eranelbaz** [comment]
> WDYT about allowing this Action to trigger the deployment of the production registry manually with workflow_dispatch?

Why do we need `workflow_dispatch` if we are always doing it to `workspace dev` and by merging isn't the workflow dispatch will make dev to run different IaCs?

**Comment by eranelbaz** [comment]
@cube2222 @Yantrio @RLRabinowitz anything else or we can push this one forward?

**Comment by RLRabinowitz** [comment]
> Why do we need workflow_dispatch if we are always doing it to workspace dev and by merging isn't the workflow dispatch will make dev to run different IaCs?

Sorry @eranelbaz I was unclear here. I meant manual deployment via `workflow_dispatch` for the **production** registry. This will allow us to know which revision is currently up in the production registry, and why

**Comment by eranelbaz** [comment]
@RLRabinowitz I think that based on this one we can create another for production deploy
I can make the changes in this PR if we would like, but IIRC I've talked to @Yantrio and we talked about first doing it only for dev

**Comment by Yantrio** [comment]
Let's create another issue for prod deployment afterwards.

I agree we need it @RLRabinowitz , but in another PR.

**Comment by Yantrio** [comment]
> Why do we need workflow_dispatch if we are always doing it to workspace dev and by merging isn't the workflow dispatch will make dev to run different IaCs?

It's better to be safe than sorry!

**Comment by eranelbaz** [comment]
> It's better to be safe than sorry!

let's add it after

File: .github/workflows/dev-deployment.yml [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+          # A role can be created by following the documentation here:
+          # https://github.com/aws-actions/configure-aws-credentials#sample-iam-oidc-cloudformation-template
+          role-to-assume: ${{ secrets.AWS_ROLE_TO_ASSUME }}
+          aws-region: ${{ secrets.AWS_REGION }}
+        continue-on-error: false # Fail the workflow if AWS configuration fails
+
+      - name: Initialize Terraform
+        if: success()
+        run: terraform init
+        continue-on-error: false
```
Comment (eranelbaz) [inline .github/workflows/dev-deployment.yml]: isn't that `false` by default?

File: .github/workflows/dev-deployment.yml [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        uses: aws-actions/configure-aws-credentials@v4
+        with:
+          # A role can be created by following the documentation here:
+          # https://github.com/aws-actions/configure-aws-credentials#sample-iam-oidc-cloudformation-template
+          role-to-assume: ${{ secrets.AWS_ROLE_TO_ASSUME }}
+          aws-region: ${{ secrets.AWS_REGION }}
+        continue-on-error: false # Fail the workflow if AWS configuration fails
+
+      - name: Initialize Terraform
+        if: success()
```
Comment (eranelbaz) [inline .github/workflows/dev-deployment.yml]: why do we need those?

File: .github/workflows/dev-deployment.yml [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        run: terraform plan -input=false -detailed-exitcode -out=project.tfplan || true
+
+      - name: Check for failure
+        if: steps.plan.outputs.exitcode != 0 && steps.plan.outputs.exitcode != 2
+        run: |
+          echo "Terraform plan failed"
+          exit 1
+
+      - name: Show the plan
+        run: terraform show project.tfplan
```
Comment (eranelbaz) [inline .github/workflows/dev-deployment.yml:60]: we don't get the plan from `plan changes?

**Comment by Yantrio** [comment]
Okay, Maybe I should clarify.

The reason I've added workflow_dispatch is because there are situations where you may need to re-deploy manually. This let's you do that. I'll be keeping it in this PR

File: .github/workflows/dev-deployment.yml [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+      - name: Generate the random password file
+        if: steps.plan.outputs.exitcode == 2
+        run: openssl rand -hex -out key.bin 64
```
Comment (eranelbaz) [inline .github/workflows/dev-deployment.yml:64]: Why do we need it?
Why not just delete this file it can be generated in memory for the apply

File: .github/workflows/dev-deployment.yml (hunk 7 above)
Comment (cube2222) [reply .github/workflows/dev-deployment.yml:60]: This is much more readable, cause it doesn't include all the refreshing logs.

File: .github/workflows/dev-deployment.yml (hunk 8 above)
Comment (cube2222) [reply .github/workflows/dev-deployment.yml:64]: > Why not just delete this file it can be generated in memory for the apply

We're doing an approval step for the plan so that we can be 100% sure what will be happening on apply. If you do apply without plan file it could do anything really, you don't know.

**Review by cube2222 — APPROVED** [review]

File: .github/workflows/dev-deployment.yml (hunk 8 above)
Comment (eranelbaz) [reply .github/workflows/dev-deployment.yml:64]: how does the approval step happens?

**Comment by RLRabinowitz** [comment]
> Let's create another issue for prod deployment afterwards.
> I agree we need it @RLRabinowitz , but in another PR.

@Yantrio @eranelbaz then let's make sure this is done in a separate PR, and maybe make sure that merging this PR doesn't automatically close the issue, as the original issue mentioned production

File: .github/workflows/dev-deployment.yml (hunk 8 above)
Comment (cube2222) [reply .github/workflows/dev-deployment.yml:64]: It's configured via the repo settings for the environment ("Applying")

File: .github/workflows/dev-deployment.yml (hunk 5 above)
Comment (Yantrio) [reply .github/workflows/dev-deployment.yml]: I'll remove these, My idea was to be explicit with this, but I was not consistent and it doesn't make sense to keep this now

File: .github/workflows/dev-deployment.yml (hunk 6 above)
Comment (Yantrio) [reply .github/workflows/dev-deployment.yml]: Similar to my other comment, I felt it was best to be explicit here, but I'll remove this for consistency sake.

**Comment by eranelbaz** [comment]
@Yantrio thank you for the Fixes
IMO Looking good
